lib/nats/io/client.rb

```diff
@@ -228,30 +230,45 @@ def initialize
       # Draining
       @drain_t = nil
 
+      # Prepare for calling connect or automatic delayed connection
+      parse_and_validate_options if uri || opts.any?
+
       # Keep track of all client instances to handle them after process forking in Ruby 3.1+
       INSTANCES[self] = self
     end
 
-    # Establishes a connection to NATS.
+    # Prepare connecting to NATS, but postpone real connection until first usage.
     def connect(uri=nil, opts={})
+      if uri || opts.any?
+        @initial_uri = uri
+        @initial_options = opts
+      end
+
+      parse_and_validate_options
+      establish_connection!
+
+      self
+    end
+
+    private def parse_and_validate_options
```

Comment:
For now I extracted this method and marked it as private in place to make diff smaller.
Probably it should go somewhere below, or maybe not.